internal/pkg/feature/cmd/feature/main.go

```diff
@@ -0,0 +1,275 @@
+		return fmt.Errorf("%s: %s", name, strings.Join(problems, "; "))
+	}
+	return nil
+}
+
+type flagValidationError struct {
+	errs []error
+}
+
+func newFlagValidationError(errs []error) *flagValidationError {
+	if len(errs) == 0 {
+		return nil
+	}
+	return &flagValidationError{errs}
+}
+
+func (e *flagValidationError) Error() string {
+	var s strings.Builder
+	s.WriteString("flag validation error: \n")
+	for _, err := range e.errs {
+		s.WriteString(err.Error())
+	}
+	return s.String()
+}
+
+func validate(flags []flagConfig) error {
+	var (
+		errs []error
+		seen = make(map[string]bool, len(flags))
+	)
+	for _, flag := range flags {
+		if err := flag.Valid(); err != nil {
+			errs = append(errs, err)
+		} else if _, repeated := seen[flag.Key]; repeated {
+			errs = append(errs, fmt.Errorf("duplicate flag key '%s'", flag.Key))
+		}
+		seen[flag.Key] = true
+	}
+	if len(errs) != 0 {
+		return newFlagValidationError(errs)
+	}
+
+	return nil
+}
+
+var argv = struct {
+	in, out *string
+	pkg     *string
+}{
+	in:  flag.String("in", "", "flag configuration path"),
+	out: flag.String("out", "", "flag generation destination path"),
+	pkg: flag.String("pkg", "feature", "package name for generated code"),
+}
+
+func main() {
+	if err := run(); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+		os.Exit(1)
+	}
+	os.Exit(0)
+}
+
+func run() error {
+	flag.Parse()
+
+	in, err := os.Open(*argv.in)
+	if err != nil {
+		return err
+	}
+	defer in.Close()
+
+	configuration, err := ioutil.ReadAll(in)
+	if err != nil {
+		return err
+	}
+
+	var flags []flagConfig
+	err = yaml.Unmarshal(configuration, &flags)
+	if err != nil {
+		return err
+	}
+	err = validate(flags)
+	if err != nil {
+		return err
+	}
+
+	t, err := template.New("flags").Funcs(templateFunctions()).Parse(tmpl)
+	if err != nil {
+		return err
+	}
+
+	out, err := os.Create(*argv.out)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	var (
+		buf  = new(bytes.Buffer)
+		vars = struct {
+			PackageName string
+			Flags       []flagConfig
+		}{
+			PackageName: *argv.pkg,
+			Flags:       flags,
+		}
+	)
+	if err := t.Execute(buf, vars); err != nil {
+		return err
+	}
+
+	raw, err := ioutil.ReadAll(buf)
+	if err != nil {
+		return err
+	}
+
+	formatted, err := format.Source(raw)
+	if err != nil {
+		return err
+	}
+
+	_, err = out.Write(formatted)
+	return err
+}
+
+func templateFunctions() template.FuncMap {
+	functions := make(template.FuncMap)
+
+	functions["quote"] = quote
+	functions["replace"] = replace
+	functions["camelcase"] = camelcase
+
+	functions["conditionalQuote"] = func(t interface{}) string {
+		switch t.(type) {
+		case string:
+			return fmt.Sprintf("%q", t)
+		default:
+			return fmt.Sprintf("%v", t)
+		}
+	}
+
+	functions["flagType"] = func(t interface{}) string {
+		switch t.(type) {
+		case bool:
+			return "BoolFlag"
+		case float64:
+			return "FloatFlag"
+		case int:
+			return "IntFlag"
+		default:
+			return "StringFlag"
+		}
```

Comment:
Yeah, again, I'm a little confused on these types. How do they fit into the feature flag?

Reply:
When we template feature flags, there's a default value and that value is inferred to be a specific type. If we see a string in the default field, we generate a `StringFlag` and this generates methods on the flag that allow us to access the feature flag with a method like `myFlag.String(ctx)` which returns the string.

This is the same with the other types and the others have equivalent names like `Int(ctx)` or `Float(ctx)`. I think booleans are the only ones that are different and use `Enabled(ctx)` instead.

This code is also copied over from the influxdb implementation.

Reply:
To point out, we very rarely, if I've ever been able to find it, use the types other than booleans. So I understand the confusion. It might be worth it for us to remove them if we find that they're causing confusion. At the moment, I think my preference would be to keep things as close to influxdb as possible.